src/Refactor/EntityWriter/ProductXmlWriter.php

```diff
@@ -0,0 +1,238 @@
+        $this->writer->endElement();
+    }
+
+    private function writeSharedVariationAttributes(): void
+    {
+        $attributes = XmlFormatter::filterEmptyValues($this->product->sharedVariationAttributes);
+
+        if (count($attributes) === 0) {
+            return;
+        }
+
+        $this->writer->startElement('attributes');
+
+        foreach ($attributes as $id) {
+            $this->writer->writeEmptyElementWithAttributes('shared-variation-attribute', [
+                'variation-attribute-id' => $id,
+                'attribute-id'           => $id,
+            ]);
+        }
+
+        $this->writer->endElement();
+    }
+
+    private function writeVariants(): void
+    {
+        $variants = XmlFormatter::filterEmptyValues($this->product->variants);
+
+        if (count($variants) === 0) {
+            return;
+        }
+
+        $this->writer->startElement('variants');
+
+        foreach ($variants as $variantId => $variantDefault) {
+            $attributes = ['product-id' => $variantId];
+
+            if ($variantDefault) {
+                $attributes['default'] = 'true';
+            }
+
+            $this->writer->writeEmptyElementWithAttributes('variant', $attributes);
+        }
+
+        $this->writer->endElement();
+    }
+
+    public function writeBundleProducts(): void
+    {
+        if ($this->product->bundleProducts === null) {
+            return;
+        }
+
+        $products = XmlFormatter::filterEmptyValues($this->product->bundleProducts);
+
+        $this->writer->startElement('bundled-products');
+
+        foreach ($products as $productId => $quantity) {
+            $this->writer->startElement('bundled-product');
+            $this->writer->writeAttribute('product-id', $productId);
+            $this->writer->writeElement('quantity', $quantity);
+            $this->writer->endElement();
+        }
+
+        $this->writer->endElement();
+    }
+
+    public function writeSetProducts(): void
+    {
+        if ($this->product->setProducts === null) {
+            return;
+        }
+
+        $products = XmlFormatter::filterEmptyValues($this->product->setProducts);
+
+        $this->writer->startElement('product-set-products');
+
+        foreach ($products as $productId) {
+            $this->writer->writeEmptyElementWithAttributes('product-set-product', ['product-id' => $productId]);
+        }
+
+        $this->writer->endElement();
+    }
+
+    public function writeClassificationCategory(): void
+    {
+        if (XmlFormatter::isEmptyValue($this->product->classificationCatalogId) && XmlFormatter::isEmptyValue($this->product->classificationCategoryId)) {
+            return;
+        }
+
+        if (! XmlFormatter::isEmptyValue($this->product->classificationCategoryId)) {
+            $this->writer->writeElementWithAttributes('classification-category', $this->product->classificationCategoryId, [
+                'catalog-id' => $this->product->classificationCatalogId,
```

Comment:
Is it worth DRYing the attribute array before the if/else?

Reply:
@WillJW I'm sure I understand what you mean here. Could you elaborate?

Reply:
@ziadoz 

```php
$attributes = ['catalog-id' => $this->product->classificationCatalogId];

if (! XmlFormatter::isEmptyValue($this->product->classificationCategoryId)) {
    $this->writer->writeElementWithAttributes('classification-category', $this->product->classificationCategoryId, $attributes);
} else {
    $this->writer->writeEmptyElementWithAttributes('classification-category', $attributes);
}
```

Reply:
@WillJW I think I'd leave it as it is tbh, at least for now.